### PR TITLE
#50990 Increase to_dict('list') performance

### LINKED ISSUE #50990: PERF: to_dict('list') method slow
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this issue exists on the main branch of pandas.

### Reproducible Example

I want to transform pandas DataFrames to dictionaries in order to hand over data as python std lib object. I had the idea of using the to_dict method. My prior experience with pandas and numpy was that my custom few liners never can compete with the provided methods performance wise. But when I noticed that the to_dict method can become a bottle neck I tried out a rather simple attempt to achieve what I wanted. I measured something like a improvement of factor 5, which is very much noticeable to me, as the transformation with to_dict can take tens of seconds for me.

I wonder if this can be improved or whether my method has some draw backs I can't anticipate right now. 

```python
import pandas as pd
import time


data_frame = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn'
                        '-data/master/iris.csv')

t0 = time.perf_counter_ns()
dict_to_dict = data_frame.to_dict('list')
t1 = time.perf_counter_ns()

values = data_frame.values
dict_custom = {}
for i, col in enumerate(data_frame.columns):
    dict_custom[col] = values[:,i].tolist()

t2 = time.perf_counter_ns()

print(f"Time with to_dict('list'):\t{(t1-t0)/10**9},\n"
      f"Time with custom method:\t{(t2-t1)/10**9}"
      )
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 2e218d10984e9919f0296931d92ea851c6a6faf5
python           : 3.8.7.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.17763
machine          : AMD64
processor        : Intel64 Family 6 Model 85 Stepping 4, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252
pandas           : 1.5.3
numpy            : 1.24.1
pytz             : 2022.7.1
dateutil         : 2.8.2
setuptools       : 65.6.3
pip              : 22.3.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

</details>

### Prior Performance

_No response_

### PR DESCRIPTION
**Pull request by taytzehao**
- [x] closes #50990 
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.1.0.rst` file if fixing a bug or adding a new feature.

`25.8±2ms       1.61±0.4ms     0.06  frame_methods.ToDict.time_to_dict_datetimelike('list')`

### CODE REVIEW

**Comment by taytzehao** [comment]
@phofl , should another issue be raised for other orients too?

File: doc/source/whatsnew/v2.1.0.rst [hunk 1]
```diff
@@ -599,6 +599,7 @@ Performance improvements
 - Performance improvement in :func:`read_orc` when reading a remote URI file path. (:issue:`51609`)
 - Performance improvement in :func:`read_parquet` and :meth:`DataFrame.to_parquet` when reading a remote file with ``engine="pyarrow"`` (:issue:`51609`)
 - Performance improvement in :func:`read_parquet` on string columns when using ``use_nullable_dtypes=True`` (:issue:`47345`)
+- Performance improvement in :func:`to_dict` on converting DataFrame to dictionary (:issue:`50990`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:602]: Can you move this to 2.2.0.rst?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -2037,6 +2037,7 @@ def to_dict(
         [defaultdict(<class 'list'>, {'col1': 1, 'col2': 0.5}),
          defaultdict(<class 'list'>, {'col1': 2, 'col2': 0.75})]
         """
+
```
Comment (mroeschke) [inline pandas/core/frame.py:2040]:
```suggestion
```

File: pandas/core/methods/to_dict.py [hunk 3]
```diff
@@ -72,6 +72,7 @@ def to_dict(
         Return a collections.abc.Mapping object representing the DataFrame.
         The resulting transformation depends on the `orient` parameter.
     """
+
```
Comment (mroeschke) [inline pandas/core/methods/to_dict.py:75]:
```suggestion
```

File: pandas/core/methods/to_dict.py [hunk 4]
```diff
@@ -106,15 +107,16 @@ def to_dict(
         return into_c((k, v.to_dict(into)) for k, v in df.items())
 
     elif orient == "list":
-        object_dtype_indices_as_set = set(box_native_indices)
+        object_dtype_indices_as_set: set[int] = set(box_native_indices)
+
```
Comment (mroeschke) [inline pandas/core/methods/to_dict.py:111]:
```suggestion
```

File: pandas/core/methods/to_dict.py [hunk 5]
```diff
@@ -106,15 +107,16 @@ def to_dict(
...
         return into_c(
             (
-                k,
-                list(map(maybe_box_native, v.tolist()))
+                col,
+                list(map(maybe_box_native, df[col].values.tolist()))
```
Comment (mroeschke) [inline pandas/core/methods/to_dict.py:115]: Is `.values` needed here? (and below)

Comment (taytzehao) [reply pandas/core/methods/to_dict.py:115]: yes, this is needed. This is to access the `np.ndarray` data type for the particular column. The `tolist()` method for `np.ndarray` is signnificantly faster that `pd.Series`

Comment (mroeschke) [reply pandas/core/methods/to_dict.py:115]: Interesting what's the performance diff between the two?

Comment (taytzehao) [reply pandas/core/methods/to_dict.py:115]: Screenshot 2023-08-31 at 2 36 04 AM

Comment (mroeschke) [reply pandas/core/methods/to_dict.py:115]: I was referring to specifically `df[col].tolist()` vs `df[col].to_numpy().tolist()`

Comment (taytzehao) [reply pandas/core/methods/to_dict.py]: This is `df[col].to_numpy().tolist()`
Screenshot 2023-08-31 at 11 13 35 PM
This is `df[col].tolist()`. `list` type orient does not change in performance
Screenshot 2023-08-31 at 11 14 07 PM
This is `df[col].values.tolist()`
Screenshot 2023-08-31 at 11 14 45 PM

File: pandas/core/methods/to_dict.py [hunk 6]
```diff
@@ -106,15 +106,15 @@ def to_dict(
...
-                k,
-                list(map(maybe_box_native, v.tolist()))
+                col,
+                list(map(maybe_box_native, df[col].to_numpy().tolist()))
                 if i in object_dtype_indices_as_set
-                else v.tolist(),
+                else df[col].to_numpy().tolist(),
             )
-            for i, (k, v) in enumerate(df.items())
+            for i, col in enumerate(df.columns)
```
Comment (mroeschke) [inline pandas/core/methods/to_dict.py:117]: Could you add a test where `to_dict("list")` is called on a DataFrame with duplicate column labels?

Comment (taytzehao) [reply pandas/core/methods/to_dict.py]: thanks for the catch, didin think about it

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @taytzehao

**Pull request merged**